inst/tests/tests.Rraw

```diff
@@ -18702,3 +18686,120 @@ test(2234.7, DT[, min(.SD), by=c(.I)], data.table(I=1L:5L, V1=c(1L, 2L, 3L, 2L,
 test(2234.8, DT[, min(.SD), by=.I%%2L], error="by.*contains .I.*supported")  # would be nice to support in future; i.e. by odd/even rows, and by=(.I+1L)%/%2L for pairs of rows; i.e. any expression of .I
 test(2234.9, DT[, min(.SD), by=somefun(.I)], error="by.*contains .I.*supported")
 
+# add na.rm argument to first/last, #4446 #4239
+DT = data.table( grp = INT(c(1,1,     2,2,2,               3,          4,4)),
+                 A =   INT(c(1,NA,    NA,2,NA,             NA,         3,3)),
+                 B =       c(NA,1,    pi,3,NA,             3,          pi,NA),
+                 C =       c("a",NA,  NA,"b",NA,           "c",        "d",NA),
+                 D =       c(NA,1i,   2i,3i,NA,            NA,         4i,NA),
+                 E =    list(NA,NA,   NA,c("a","b"),NULL,  list(1:2),  3:4,NA))
+# to encourage adding cases to DT above (some NAs, all NAs, NAs beginning and/or end within group, types, etc) we
+# define helper functions FNNA and LNNA here to construct the correct result to test against. Explicitly coding
+# all the results instead may reveal mistakes and be good but then might be inflexible to adding more cases.
+FNNA = function(x) {  # First Not NA
+  nna = which(!is.na(x) & !(is.list(x) & sapply(x,is.null)))
+  if (length(nna)) x[nna[1L]]          else if (is.list(x)) list(NA) else x[NA_integer_]
+}
+LNNA = function(x) {  # Last Not NA
+  nna = which(!is.na(x) & !(is.list(x) & sapply(x,is.null)))
+  if (length(nna)) x[nna[length(nna)]] else if (is.list(x)) list(NA) else x[NA_integer_]
+}
+test_no = 2235.0
+for (col in setdiff(names(DT), "grp")) {
+  # test vector input to first/last outside of DT[] query
+  test(test_no+.01, first(DT[[col]], na.rm=TRUE), ans<-FNNA(DT[[col]]))
+  test(test_no+.02, first(DT[[col]], na.rm="row"), ans)   # "row" same as TRUE for vectors
+  test(test_no+.03, last(DT[[col]], na.rm=TRUE), ans<-LNNA(DT[[col]]))
+  test(test_no+.04, last(DT[[col]], na.rm="row"), ans)
+  
+  # one function by group with and without .() wrapper
+  test(test_no+.05, EVAL("DT[, .(first(",col,",na.rm=TRUE)), by=grp, verbose=TRUE]"),
+               ans<-EVAL("DT[, .(FNNA(",col,")),             by=grp]"), output="GForce optimized.*gfirst")
+  if (col!="E")  # subets of list columns need to be done within .() otherwise j's result looks like multiple columns
+    test(test_no+.06, EVAL("DT[, first(",col,",na.rm=TRUE),  by=grp, verbose=TRUE]"),
+                      ans,                                              output="GForce optimized.*gfirst")
+  test(test_no+.07, EVAL("DT[, .(last(",col,",na.rm=TRUE)),  by=grp, verbose=TRUE]"),
+               ans<-EVAL("DT[, .(LNNA(",col,")),             by=grp]"), output="GForce optimized.*glast")
+  if (col!="E")
```

Comment:
add a comment about why the special handling of E (I guess it's because of `.()` vs `.(list())` subtlety?

Reply:
there's already comment just above: `# subsets of list columns need to be done within .() otherwise j's result looks like multiple columns`.  Added `# see comment above why !="E"` here.

Reply:
thanks... missed that. extra comment can't hurt